File: daily/Backtracking/40-Combination-sum-ii.go
package daily

import "sort"

func combinationSum2(candidates []int, target int) [][]int {
	sort.Ints(candidates)
	res := [][]int{}
	path := []int{}

	var backtrack func(start, remaining int)
	backtrack = func(start, remaining int) {
		if remaining == 0 {
			combo := make([]int, len(path))
			copy(combo, path)
			res = append(res, combo)
			return
		}

		for i := start; i < len(candidates); i++ {
			if candidates[i] > remaining {
				return
			}
			if i > start && candidates[i-1] == candidates[i] {
				continue
			}
			path = append(path, candidates[i])
			backtrack(i+1, remaining-candidates[i])
			path = path[:len(path)-1]
		}
	}
	backtrack(0, target)
	return res
}

File: daily/Backtracking/40-Combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        results = []
        candidates.sort()
        def backtrack(start, path, remaining):
            if remaining == 0:
                results.append(path[:])
                return
            for i in range(start, len(candidates)):
                if candidates[i] > remaining:
                    return
                if i > start and candidates[i - 1] == candidates[i]:
                    continue
                path.append(candidates[i])
                backtrack(i + 1, path, remaining - candidates[i])
                path.pop()
        backtrack(0, [], target)
        return results